File: Recursion/GFG/Medium/BinaryStrings/BinaryStrings.cpp
// https://www.geeksforgeeks.org/problems/generate-all-binary-strings/
#include<bits/stdc++.h>
using namespace std;
void generate(int num, string str, vector<string> &combos)
{
    if (str.length() == num)
    {
        combos.push_back(str);
        return;
    }
    generate(num, str + "0", combos);
    if (str.length() == 0 || str[str.length() - 1] != '1')
    {
        generate(num, str + "1", combos);
    }
}

vector<string> generateBinaryStrings(int num)
{
    vector<string> ans;
    generate(num, "", ans);
    return ans;
}
int main(void) {
    vector<string> v = generateBinaryStrings(5);
    for(string ele: v) {
        cout << ele << " ";
    }
    cout << endl;
    return 0;
}

File: Recursion/GFG/Medium/BinaryStrings/BinaryStrings.java
// https://www.geeksforgeeks.org/problems/generate-all-binary-strings/
import java.util.ArrayList;
import java.util.List;

public class BinaryStrings {
    static void generate(int n, String s, List<String> combo) {
        if (s.length() == n) {
            combo.add(s);
            return;
        }
        generate(n, s + "0", combo);
        if (s.equals("") || s.charAt(s.length() - 1) != '1') {
            generate(n, s + "1", combo);
        }
    }

    public static List<String> generateBinaryStrings(int n) {
        List<String> ans = new ArrayList<>();
        generate(n, "", ans);
        return ans;
    }

    public static void main(String[] args) {
        List<String> str = generateBinaryStrings(5);
        System.out.println(str.toString());
    }
}
